### PR TITLE
[MRG] ENH add explained_variance to PCA

### LINKED ISSUE #2341: Explained variance of PCs
Is there an easy way to get the percentage of explained variance for each of the principal components?

### PR DESCRIPTION
**Pull request by teonlamont**
closes #2341. this is written to return an optional `explained_variance`. I am not sure if this is pythonic. The other option would be to make a separate function but then it wouldn't be DRY.

### CODE REVIEW

**Comment by jasmainak** [comment]
@teonlamont : I was thinking more in terms of adding an attributed called `explained_variance_` to the proj object instead of returning it in every function. More along the lines of sklearn API ...

**Comment by teonlamont** [comment]
i'm fine with that. that actually makes more sense ;)

File: mne/proj.py [hunk 1]
```diff
@@ -97,15 +98,23 @@ def _compute_proj(data, info, n_grad, n_mag, n_eeg, desc_prefix, verbose=None):
...
             continue
         data_ind = data[ind][:, ind]
-        U = linalg.svd(data_ind, full_matrices=False,
-                       overwrite_a=True)[0][:, :n]
-        for k, u in enumerate(U.T):
+        # data is the covariance matrix: U * S**2 * Ut
+        U, Sexp2, _ = linalg.svd(data_ind, full_matrices=False,
+                                 overwrite_a=True)
+        U = U[:, :n]
+        exp_var_ = Sexp2 / n_epochs
```
Comment (agramfort) [inline mne/proj.py]: n_epochs is not needed as you divide by the sum below.

how about just logging the explained variance and not storing it as attribute?

note that the attribute will not be saved to disk.

Comment (Eric89GXL) [inline mne/proj.py]: Would it break the C tools if we added a constant to save the value? It does seem like it could be useful to save it...

Comment (teonlamont) [inline mne/proj.py]: +1 on keeping it if possible.

Comment (agramfort) [inline mne/proj.py]: fair enough but it needs a bit more work then :)

File: mne/proj.py [hunk 2]
```diff
@@ -97,15 +97,21 @@ def _compute_proj(data, info, n_grad, n_mag, n_eeg, desc_prefix, verbose=None):
...
+        exp_var_ = exp_var_[:n]
+        for k, temp in enumerate(zip(U.T, exp_var_)):
+            u, var = temp
             proj_data = dict(col_names=names, row_names=None,
                              data=u[np.newaxis, :], nrow=1, ncol=u.size)
             this_desc = "%s-%s-PCA-%02d" % (desc, desc_prefix, k + 1)
             logger.info("Adding projection: %s" % this_desc)
             proj = Projection(active=False, data=proj_data,
                               desc=this_desc, kind=1)
+            proj.explained_variance_ = var
```
Comment (Eric89GXL) [inline mne/proj.py]: Why put the trailing `_` on the variable name? I know `sklearn` does it, but we don't typically anywhere other than `sklearn`-style classes AFAIK.

**Comment by Eric89GXL** [comment]
Actually instead of creating a new constant, maybe we can just reuse `mne_ica_pca_explained_var   3605 ? "PCA explained variance"` from (mne-fiff-constants) [https://github.com/mne-tools/mne-fiff-constants]? Maybe we could even just rename it to a more generic `mne_pca_explained_variance` (with the same value) and use it both places. Then we can add it to our projectors when we save and load them. We'd want to check that the C tools could still read it, though.

File: mne/proj.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/proj.py]: ...and it should probably be passed to the constructor, too.

Comment (teonlamont) [inline mne/proj.py]: I was also thinking that would be the proper way to go.

**Comment by teonlamont** [comment]
> Actually instead of creating a new constant, maybe we can just reuse mne_ica_pca_explained_var 3605 ? "PCA explained variance" from mne-fiff-constants?

that sounds like the way to go. I guess for older `proj` files without this file, we would return `None`?

**Comment by Eric89GXL** [comment]
Yeah. If the value is `None` (default for constructor), don't write the value. If non-None, write the value.

**Comment by Eric89GXL** [comment]
...but before you work on that, get me those head point .txt files please :)

**Comment by teonlamont** [comment]
the errors in the travis built is due to the gamma inverse map. ready for review

**Comment by Eric89GXL** [comment]
I restarted the builds, in theory they should come back okay

**Comment by Eric89GXL** [comment]
LGTM. Have you tried loading the projectors saved this way with the MNE-C tools? You can do it by creating some on raw data in mne-python, saving them, and going into mne_browse_raw and trying to apply them to the loaded data.

File: mne/proj.py [hunk 3]
```diff
@@ -97,15 +97,20 @@ def _compute_proj(data, info, n_grad, n_mag, n_eeg, desc_prefix, verbose=None):
...
-        U = linalg.svd(data_ind, full_matrices=False,
-                       overwrite_a=True)[0][:, :n]
-        for k, u in enumerate(U.T):
+        # data is the covariance matrix: U * S**2 * Ut
+        U, Sexp2, _ = linalg.svd(data_ind, full_matrices=False,
+                                 overwrite_a=True)
+        U = U[:, :n]
+        exp_var = Sexp2 / Sexp2.sum()
+        exp_var = exp_var[:n]
+        for k, temp in enumerate(zip(U.T, exp_var)):
```
Comment (agramfort) [inline mne/proj.py]: for k, temp in enumerate(zip(U.T, exp_var)):

->

for k, (u, var) in enumerate(zip(U.T, exp_var)):

File: mne/tests/test_proj.py [hunk 4]
```diff
@@ -122,6 +122,8 @@ def test_compute_proj_epochs():
                 p2_data = p2_data[:, mask]
             corr = np.corrcoef(p1_data, p2_data)[0, 1]
             assert_array_almost_equal(corr, 1.0, 5)
+            if p2['explained_var']:
+                assert_true(p1['explained_var'] == p2['explained_var'])
```
Comment (agramfort) [inline mne/tests/test_proj.py]: can you check that the value is good?

Comment (Eric89GXL) [inline mne/tests/test_proj.py]: like between 0 and 1?

I guess we could also add tests to our `preprocessing` module for EOG/ECG that the values are at least 0.1 or something.

Comment (agramfort) [inline mne/tests/test_proj.py]: >
> like between 0 and 1?
>
yes and maybe closer to 1

> I guess we could also add tests to our preprocessing module for EOG/ECG
> that the values are at least 0.1 or something.
>
exactly

File: mne/proj.py (hunk 3 above)
Comment (teonlamont) [inline mne/proj.py]: nice! didn't know that one ;)

**Comment by teonlamont** [comment]
just noticed two things with the comparison: 1. the saved file stores it as the `explained_var` as a `float32` whereas the generated one is `float64`. 2. the saved file is loaded as an array and the generated one is float.

**Comment by Eric89GXL** [comment]
Both of those things actually true of almost all our reading functions (float64 gets cast to float32; everything ends up being array). If you really want you can do `float(val)` when reading to ensure it's a singleton.

**Comment by teonlamont** [comment]
oh ok, i don't know the test passed then. would it be better to do an assert_array_almost_equal for the comparison. I am also trying to do the mne_browse_raw. I am having some x11 problems (I rarely use the mne-c functions)

**Comment by Eric89GXL** [comment]
you can use mne_process_raw probably too and avoid X11 altogether

**Comment by teonlamont** [comment]
ok, that worked. it didn't break the binary commands however upon saying the projectors with the `explained_var`, this is not written into the binary actions (understandable) so the saved file has the entry `explained_var` but it is empty.

**Comment by Eric89GXL** [comment]
Excellent. I guess if you add the tests for explained_var some reasonable amount greater than zero, especially in our `preprocessing` tests, it should be good to go.

**Comment by Eric89GXL** [comment]
... I think you meant to put `[ci skip]` in that commit message. I'll kill the builds :)

**Comment by teonlamont** [comment]
lol. i couldn't remember what it was ;) what is the test I should add to the `preprocessing`, just assert that `proj['explained_var'] > 0`?

**Comment by Eric89GXL** [comment]
Nah something bigger preferably. See how high our `EOG` and `ECG` vectors do in testing. Hopefully at least the first component is pretty big > 0.2 at least. If it's 0.05... it's not great. So I'd say make sure they're above a reasonable lower bound and less than or equal to one -- here "reasonable" I guess means the biggest nice looking number you can find such that `x > lower_bound`.

**Comment by Eric89GXL** [comment]
Setting the threshold that way will allow us to see how we're doing in terms of explained variance, and make sure we don't introduce regressions in our SSP code.

**Comment by Eric89GXL** [comment]
I've used this sort of highest-nice-bound technique in the Maxwell filtering tests to set the minimum and median bound check in a custom `assert_meg_snr` function, e.g.:

https://github.com/mne-tools/mne-python/blob/master/mne/preprocessing/tests/test_maxwell.py#L219

It seems to be working well so far to spot regressions.

**Comment by teonlamont** [comment]
Some components are explaining a lot, some aren't. 
```python
In [8]: ssp_eog = [proj for proj in projs if proj['desc'].startswith('EOG')]
In [9]: proj['explained_var'] for proj in ssp_eog]
Out[9]: [0.12477206325547088, 0.072218770574639407,
         0.36892777667722315, 0.15824328731324769,
         0.91739351958113013, 0.039492122694309253]
```

**Comment by Eric89GXL** [comment]
I'd look at the first component of each channel type probably. I bet that's what the first column is already. And that's about what I'd expect: EEG (last) is highest, mag (second) is next, followed by grad (not awesome). So I'd take that first column and use it (with some comment about why) and ensure each is greater than 0.1, 0.3, and 0.9 respectively, and less than or equal to one.

**Comment by Eric89GXL** [comment]
(and I'd do something similar for ECG, too)

**Comment by teonlamont** [comment]
got it. and nice, you were spot on. 
```python
In [10]: {proj['desc']: proj['explained_var'] for proj in ssp_eog}
Out[10]: {'EOG-eeg-998--0.200-5.000-PCA-01': 0.91739351958113013,
          'EOG-eeg-998--0.200-5.000-PCA-02': 0.039492122694309253,
          'EOG-axial-998--0.200-5.000-PCA-01': 0.36892777667722315,
          'EOG-axial-998--0.200-5.000-PCA-02': 0.15824328731324769,
          'EOG-planar-998--0.200-5.000-PCA-01': 0.12477206325547088, 
          'EOG-planar-998--0.200-5.000-PCA-02': 0.072218770574639407}

**Comment by jasmainak** [comment]
@teonlamont can you update a test to read and write the `proj` which has explained var in it?

**Comment by teonlamont** [comment]
@jasmainak that's taken care of in https://github.com/mne-tools/mne-python/pull/2703/files#diff-9c603f5cb5c525cc07ef4656938182c3R102

**Comment by jasmainak** [comment]
> @jasmainak that's taken care of in https://github.com/mne-tools/mne-python/pull/2703/files#diff-9c603f5cb5c525cc07ef4656938182c3R102

excellent!

**Comment by teonlamont** [comment]
I think it's ready

**Comment by agramfort** [comment]
thanks @teonlamont
